### PR TITLE
Master ecommerce hide price shda

### PR DESCRIPTION
**Pull request by donnie-blurry**
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/website_sale/models/website.py [hunk 1]
```diff
@@ -78,6 +78,9 @@ def _get_product_sort_mapping():
         ],
         default='optional')
 
+    hide_add_to_cart = fields.Boolean(string="Hide 'Add To Cart' when price = 0")
+    contact_us_button_url = fields.Char(string="Contact Us Button URL")
```
Comment (Williambraecky) [inline addons/website_sale/models/website.py:82]: Maybe we can add a default to `/contactus` ?

File: addons/website_sale/static/src/js/variant_mixin.js [hunk 2]
```diff
@@ -39,7 +39,18 @@ VariantMixin._onChangeCombination = function (ev, $parent, combination) {
...
         $product.trigger('view_item_event', combination['product_tracking_info']);
     }
-
+    const addToCart = $parent.find('#o_wsale_cta_wrapper');
+    const contactUsButton = $parent.find('#contact_us_wrapper');
+    const productPrice = $parent.find('.product_price');
+    if (!combination.price) {
+        productPrice.removeClass('d-inline-block').addClass('d-none');
+        addToCart.removeClass('d-flex').addClass('d-none');
+        contactUsButton.removeClass('d-none').addClass('d-flex');
```
Comment (Williambraecky) [inline addons/website_sale/static/src/js/variant_mixin.js:48]: I'm never sure about jQuery things, does this work whenever the option is disabled?
Will it now throw an error?

**Review by Williambraecky — CHANGES_REQUESTED** [review]
Nice work 😄 

For security we also need to prevent an user from adding a product that has a price of 0 to the cart on the python side I believe.

File: addons/website_sale/static/src/js/variant_mixin.js (hunk 2 above)
Comment (donnie-blurry) [reply addons/website_sale/static/src/js/variant_mixin.js:48]: I'm also not sure about jQuery things, I'm very new to them
To my understanding the option just controls the presence of contact us button.
The add to cart button and price will be hidden anyway.

Comment (Williambraecky) [reply addons/website_sale/static/src/js/variant_mixin.js:48]: What I mean is if contactUsButton is not found, will removeClass not throw an error?

File: addons/website_sale/models/website.py (hunk 1 above)
Comment (donnie-blurry) [reply addons/website_sale/models/website.py]: Done

File: addons/website_sale/views/templates.xml [hunk 3]
```diff
@@ -213,7 +214,7 @@
         <xpath expr="//div[hasclass('o_wsale_product_btn')]" position="inside">
             <t t-set="product_variant_id" t-value="product._get_first_possible_variant_id()"/>
             <input name="product_id" t-att-value="product_variant_id" type="hidden"/>
-            <t t-if="product_variant_id">
+            <t t-if="product_variant_id and template_price_vals['price_reduce']">
```
Comment (donnie-blurry) [inline addons/website_sale/views/templates.xml:217]: @Williambraecky I'm not sure why you did this, but I did not change it

File: addons/website_sale/models/sale_order.py [hunk 4]
```diff
@@ -105,6 +105,9 @@ def _cart_update(self, product_id, line_id=None, add_qty=0, set_qty=0, **kwargs)
         if not product:
             raise UserError(_("The given product does not exist therefore it cannot be added to cart."))
 
+        if product.lst_price == 0:
```
Comment (Williambraecky) [inline addons/website_sale/models/sale_order.py:108]: Only if the website option is enabled :)

**Review by Williambraecky — COMMENTED** [review]
Small review 😄 
Thanks

File: addons/website_sale/models/product_product.py [hunk 5]
```diff
@@ -88,7 +88,9 @@ def _is_sold_out(self):
         return combination_info['product_type'] == 'product' and combination_info['free_qty'] <= 0
 
     def _website_show_quick_add(self):
-        return self.sale_ok
+        website = self.env['website'].get_current_website()
+        prevent_zero_price_sale = website and website.prevent_zero_price_sale
+        return self.sale_ok and (self._get_contextual_price() or not prevent_zero_price_sale)
```
Comment (Williambraecky) [inline addons/website_sale/models/product_product.py:93]: FYI `get_current_website` will always return a record, empty or not.
Also FYI accessing a field on an empty recordset will yield a falsy value, as such your condition here is not really necessary.
```suggestion
        website = self.env['website'].get_current_website()
        return self.sale_ok and (self._get_contextual_price() or not website.prevent_zero_price_sale)
```

File: addons/website_sale/models/product_template.py [hunk 6]
```diff
@@ -515,4 +519,6 @@ def _get_contextual_pricelist(self):
         return pricelist
 
     def _website_show_quick_add(self):
-        return self.sale_ok
+        website = self.env['website'].get_current_website()
+        prevent_zero_price_sale = website and website.prevent_zero_price_sale
+        return self.sale_ok and (self._get_contextual_price() or not prevent_zero_price_sale)
```
Comment (Williambraecky) [inline addons/website_sale/models/product_template.py:524]: Same as in `product_product.py`

**Review by Williambraecky — COMMENTED** [review]
Small review! Thanks!

**Review by Williambraecky — COMMENTED** [review]
Hello 👋 
Thanks for your work, I found one issue I would like you to look into :)

File: addons/website_sale/models/product_product.py [hunk 7]
```diff
@@ -88,7 +88,8 @@ def _is_sold_out(self):
         return combination_info['product_type'] == 'product' and combination_info['free_qty'] <= 0
 
     def _website_show_quick_add(self):
-        return self.sale_ok
+        website = self.env['website'].get_current_website()
+        return self.sale_ok and (self._get_contextual_price() or not website.prevent_zero_price_sale)
```
[the remark this answers is not in the document]
Comment (Williambraecky) [reply addons/website_sale/models/product_product.py:92]: Actually it won't be called that much so it's ok, still want to swap the condition though.

File: addons/website_sale/data/product_snippet_template_data.xml [hunk 8]
```diff
@@ -287,10 +287,17 @@
         </template>
 
         <template id="price_dynamic_filter_template_product_product" name="Dynamic Product Filter Price">
-            <span t-esc="record._get_contextual_price()" class="fw-bold"
-                  t-options="{'widget': 'monetary', 'display_currency': website.currency_id}"/>
-            <del t-if="data.get('has_discounted_price')" class="text-danger ms-1 h6" style="white-space: nowrap;"
-                 t-esc="data['list_price']" t-options="{'widget': 'monetary', 'display_currency': website.currency_id}"/>
+            <t t-set="record_price" t-value="record._get_contextual_price()"/>
+            <div t-if="not website.prevent_zero_price_sale or record_price">
```
Comment (Williambraecky) [inline addons/website_sale/data/product_snippet_template_data.xml:291]: FYI for purely logical stuff you can use the `t` tag
```suggestion
            <t t-if="not website.prevent_zero_price_sale or record_price">
```

**Comment by Williambraecky** [comment]
Thanks for your work!

robodoo r+

**Pull request closed without merging**